packages/component/src/Middleware/Activity/createCoreMiddleware.js

```diff
@@ -25,37 +25,43 @@ export default function createCoreMiddleware() {
         // Do not show empty bubbles (no text and attachments, and not "typing")
         !(text || attachments.length)
       ) {
-        return RETURN_FALSE;
+        return false;
       }
     } else if (type === 'typing' && activity.from.role === 'user') {
       // Do not show typing by oneself
-      return RETURN_FALSE;
+      return false;
     }
 
     if (type === 'message' || type === 'typing') {
       if (type === 'message' && (activity.attachments || []).length > 1 && activity.attachmentLayout === 'carousel') {
-        // The following line is not a React functional component, it's a middleware function.
-        // Note that "children" is not a props, but first argument.
-        const CarouselActivity = children => (
-          <CarouselLayout activity={activity} nextVisibleActivity={nextVisibleActivity}>
-            {children}
-          </CarouselLayout>
-        );
-
-        return CarouselActivity;
+        // The following line is not a React functional component, it's a render function called by useCreateActivityRenderer() hook.
+        // The function signature need to be compatible with older version of activity middleware, which was:
+        //
+        // renderActivity(
+        //   renderAttachment: ({ activity, attachment }) => React.Element
+        // ) => React.Element
+
+        return function renderCarouselLayout(renderAttachment, props) {
+          typeof props === 'undefined' &&
+            console.warn(
+              'botframework-webchat: One or more arguments were missing after passing through the activity middleware. Please check your custom activity middleware to make sure it passthrough all arguments.'
+            );
+
+          return <CarouselLayout activity={activity} renderAttachment={renderAttachment} {...props} />;
+        };
       }
 
-      // The following line is not a React functional component, it's a middleware function.
-      // Note that "children" is not a props, but first argument.
-      const StackedActivity = children => (
-        <StackedLayout activity={activity} nextVisibleActivity={nextVisibleActivity}>
-          {children}
-        </StackedLayout>
-      );
+      // The following line is not a React functional component, it's a render function called by useCreateActivityRenderer() hook.
+      return function renderStackedLayout(renderAttachment, props) {
+        typeof props === 'undefined' &&
```

Comment:
I think just "props === undefined" would work here; there's no need for the typeof.

Reply:
Since we don't control the hosting environment, if someone assign `window.undefined` to `1`, then, we use `props === undefined`, it will break our code.

https://stackoverflow.com/questions/5663277/what-is-the-difference-between-undefined-and-undefined

Reply:
Weird edge case, but I can understand why you'd want to do that. Okay.